Issue 225

vqd8a opened the pull request:
Address #225

vqd8a commented:
Added arith traits overload for vector complex types in [KokkosBatched_Vector.hpp](https://github.com/kokkos/kokkos-kernels/blob/issue-225/src/batched/KokkosBatched_Vector.hpp) 
Modified [Test_Batched_VectorArithmatic.hpp](https://github.com/kokkos/kokkos-kernels/blob/issue-225/unit_test/batched/Test_Batched_VectorArithmatic.hpp) respectively.
Still leave RealPart() in [KokkosBatched_Vector.hpp](https://github.com/kokkos/kokkos-kernels/blob/issue-225/src/batched/KokkosBatched_Vector.hpp) since I am not sure where it is called.

vqd8a commented:
Thanks, @kyungjoo-kim.
Couldn't these be taken care by imag() in ArithTraits? Why did we need ImagPart for these cases?
```C++
template<typename T> KOKKOS_FORCEINLINE_FUNCTION typename Kokkos::Details::ArithTraits<T>::mag_type ImagPart(const T &val) { return 0; }
template<typename T> KOKKOS_FORCEINLINE_FUNCTION typename Kokkos::Details::ArithTraits<Kokkos::complex<T> >::mag_type ImagPart(const Kokkos::complex<T> &val) { return val.imag(); }
```
I will change RealPart and ImagPart used in other routines too.

kyungjoo-kim commented:
if you added traits, it is not necessary. they are just quick solution when I wrote first time.

vqd8a commented:
Changed RealPart and ImagPart used in other routines to real() and imag().

vqd8a commented:
Thanks @mhoemmen . The code is fixed.

kyungjoo-kim commented:
@vqd8a Yhis needs to wait until #232 is merged first.

vqd8a commented:
Thanks, @kyungjoo-kim . Could you please help answer @mhoemmen 's questions about MagnitudeScalarType and ValueScalarType?

kyungjoo-kim commented:
@vqd8a They are probably a bi-product when I develop the code. It is a quick solution without thinking too much ;). I don't think that I use it in the code but we can change the code as @mhoemmen suggests.

mhoemmen commented:
@kyungjoo-kim PURGE THEM :D

vqd8a commented:
Since ValueScalarType is not used elsewhere, I just removed it.
As @mhoemmen suggested, I also changed MagnitudeScalarType to ArithTraits<T>::mag_type. Other routines that used MagnitudeScalarType were also changed accordingly.

kyungjoo-kim commented on src/batched/KokkosBatched_LU_Decl.hpp:
```diff
@@ -16,7 +16,7 @@ namespace KokkosBatched {
       invoke(const AViewType &A,
-             const typename MagnitudeScalarType<typename AViewType::non_const_value_type>::type tiny = 0);
+             const typename Kokkos::Details::ArithTraits<typename AViewType::non_const_value_type>::mag_type tiny = 0);
     };       
```
This is wrong. In the batched routines, input value_type can be two -  double or SIMD<double> type. When SIMD<double> type is used, your modification returns SIMD<mag_type<double> > but the previous routine MagnitudeScalarType returns double type.

vqd8a replied:
Should we keep ```MagnitudeScalarType```?

kyungjoo-kim replied:
You have two options: keep MagnitudeScalarType or using the ValueScalarType
```
typename Kokkos::Details::ArithTraits<typename ValueScalarType<typename AViewType::non_const_value_type>::type>::mag_type
```

vqd8a replied:
I think we should go back to MagnitudeScalarType

vqd8a commented:
@kyungjoo-kim Please review these changes.

kyungjoo-kim commented:
BTW, did you check the code compiles and pass all the unit tests ?

vqd8a commented:
thanks @kyungjoo-kim . Yes, it passed all the unit tests.

kyungjoo-kim commented:
thank you for fixing the code.

crtrott commented:
Passed tests other than the clang 6 error in SPGEMM_KK_MEMORY:
```
#######################################################
PASSED TESTS
#######################################################
clang-3.9.0-Pthread_Serial-release build_time=225 run_time=1702
cuda-9.1-Cuda_OpenMP-release build_time=619 run_time=1091
gcc-4.8.4-OpenMP-release build_time=138 run_time=317
gcc-4.8.4-Pthread-release build_time=125 run_time=879
gcc-5.3.0-Serial-release build_time=153 run_time=763
intel-16.0.1-OpenMP-release build_time=403 run_time=374
#######################################################
FAILED TESTS
#######################################################
clang-6.0-Cuda_Pthread-release (test failed)
```

mhoemmen commented on src/batched/KokkosBatched_Vector.hpp:
```diff
@@ -138,13 +138,13 @@ namespace Kokkos {
 	
     template<typename T, int l>
-    class ArithTraits<Vector<SIMD<Kokkos::complex<T>>,l> > { 
+    class ArithTraits<KokkosBatched::Experimental::Vector<KokkosBatched::Experimental::SIMD<Kokkos::complex<T>>,l> > { 
```
No love for my template template, alas...

crtrott replied:
Uhm that didn't compile without the change ...

vqd8a replied:
when I looked at the old code RealPart and ImagPart, I saw they were only for SIMD. That's why I did not try @mhoemmen 's solution template template.
Now, I have just tried it and it works. Should I create a new PR for this?

mhoemmen replied:
@vqd8a I don't mean to tell you what to do :-)  If it needs to work for AVX, etc. as well as SIMD, then my approach would give you a generic solution.

vqd8a replied:
@crtrott @kyungjoo-kim should we need this generic solution?

mhoemmen replied:
(You don't _need_ it, but you do need ArithTraits, etc. to work for `AVX<T>` and any other `SimdImplementation<T>` you want to support, as well as `SIMD<T>`.  However that happens is up to you :-) .)

kyungjoo-kim commented:
right now I unified avx into simd. but I still open the possibility that someone else wants to try this kind of structure with a different tag that can guide to a different instruction set e.g., ARM. I think that the current solution is fine and let's not worry about things not happend yet.
